zcash_client_sqlite/src/wallet.rs

```diff
@@ -2020,9 +2022,29 @@ pub(crate) fn truncate_to_height<P: consensus::Parameters>(
         named_params![":new_end_height": u32::from(block_height + 1)],
     )?;
 
-    // If we're removing scanned blocks, we need to truncate the note commitment tree, un-mine
-    // transactions, and remove received transparent outputs and affected block records from the
-    // database.
+    // Mark transparent utxos as un-mined. Since the TXO is now not mined, it would ideally be
+    // considered to have been returned to the mempool; it _might_ be spendable in this state, but
+    // we must also set its max_observed_unspent_height field to NULL because the transaction may
+    // be rendered entirely invalid by a reorg that alters anchor(s) used in constructing shielded
+    // spends in the transaction.
+    conn.execute(
+        "UPDATE transparent_received_outputs
+         SET max_observed_unspent_height = NULL
+         WHERE max_observed_unspent_height > :height",
+        named_params![":height": u32::from(block_height)],
+    )?;
+
+    // Un-mine transactions. This must be done outside of the last_scanned_height check because
+    // transaction entries may be created as a consequence of receiving transparent TXOs.
+    conn.execute(
+        "UPDATE transactions
+         SET block = NULL, mined_height = NULL, tx_index = NULL
+         WHERE block > ?",
```

Comment:
Hm, yeah, I guess that given the constraints on `mined_height` it's okay to use it in this fashion. This makes it a bit more load-bearing than I had originally anticipated.

Reply:
Wait a minute: you're reading documentation about sort order, not about binary operations involving NULL:
```
sqlite> select NULL < 1;

sqlite> select 2 < 1;
0
sqlite> select 1 < 2;
1
```

Reply:
After a long Slack thread discussion, we determined that the SQLite documentation I linked above is Just Wrong, because it is _very clearly_ documenting binary operations, _not_ sort order. But the binary operations actually behave like mathematical operations (if an argument is NULL, the result is NULL), despite the separate mathematical operations section very specifically limiting itself to not apply to binary operations.

Reply:
None of this materially affects my bug reports; it just means that instead of `height = NULL` being treated as "infinite confirmations", any row matching `height = NULL` is ignored completely. The latter behaviour is incorrect when `max_height = chain_height + 1` (targeting the mempool height), so it still comes down to "unmined transactions are not correctly treated as being in the mempool by the SQL queries".